### PR TITLE
Enforce constraint C1128 for DO CONCURRENT locality-spec's

### PR DESCRIPTION
**Pull request by psteinfeld**
These changes implement most of the requirements for C1128, which says: "A
variable-name that appears in a LOCAL or LOCAL_INIT locality-spec shall not
have the ALLOCATABLE; INTENT (IN); or OPTIONAL attribute; shall not be of
finalizable type; shall not be a nonpointer polymorphic dummy argument; and
shall not be a coarray or an assumed-size array.  A variable-name that is not
permitted to appear in a variable definition context shall not appear in a
LOCAL or LOCAL_INIT locality-spec."

The changes do not implement the checking required to determine whether a
variable can appear in a "variable definition context".

Here's a summary of the changes:
 - I created the function 'PassesLocalityChecks()' to enforce C1128 along with
   C1124, C1125, and C1126.
 - I cleaned up the code to check if a type or symbol is a coarray.
 - I added functions to tools.[h,cc] to test if a symbol is OPTIONAL, INTENT
   IN, finalizable, a coarray, or an assumed size array.  Should these be
   member functions of the type "Symbol"?
 - Since I changed one of the locality related error messages, I needed to
   change the test resolve35.f90.
 - I added the test resolve55.f90 to test all of the checks implemented in this
   update.

### CODE REVIEW

File: lib/semantics/resolve-names.cc [hunk 1]
```diff
@@ -3672,8 +3676,62 @@ bool DeclarationVisitor::HandleUnrestrictedSpecificIntrinsicFunction(
...
+        "Finalizable variable '%s' not allowed in a locality-spec"_err_en_US);
+    return false;
+  }
+  if (IsCoarray(symbol)) {  // C1128
+    SayWithDecl(
+        name, symbol, "Coarray '%s' not allowed in a locality-spec"_err_en_US);
+    return false;
+  }
+  const DeclTypeSpec *type{symbol.GetType()};
+  if (type) {
```
Comment (ichoyjx) [inline lib/semantics/resolve-names.cc]: Maybe simply `if (const DeclTypeSpec *type{symbol.GetType()}) {` ?

File: lib/semantics/tools.cc [hunk 2]
```diff
@@ -387,4 +385,33 @@ const Symbol *FindUltimateComponent(const DerivedTypeSpec &derivedTypeSpec,
   return nullptr;
 }
 
+bool IsFinalizable(const Symbol &symbol) {
+  const DeclTypeSpec *type{symbol.GetType()};
+  if (type) {
+    const DerivedTypeSpec *derived{type->AsDerived()};
+    if (derived) {
+      const Scope *scope{derived->scope()};
+      if (scope) {
```
Comment (ichoyjx) [inline lib/semantics/tools.cc]: `type`, `derived`, and `scope`, same as above.

File: lib/semantics/resolve-names.cc [hunk 3]
```diff
@@ -3672,8 +3676,62 @@ bool DeclarationVisitor::HandleUnrestrictedSpecificIntrinsicFunction(
...
+    if (type->IsPolymorphic() && symbol.IsDummy() &&
+        (!IsPointer(symbol))) {  // C1128
```
Comment (klausler) [inline lib/semantics/resolve-names.cc]: Surplus (parentheses)

File: lib/semantics/tools.cc (hunk 2 above)
Comment (psteinfeld) [reply lib/semantics/tools.cc]: Thanks.  I'll change that.

File: lib/semantics/resolve-names.cc (hunk 3 above)
Comment (psteinfeld) [reply lib/semantics/resolve-names.cc]: Thanks.  I'll change that.

File: lib/semantics/resolve-names.cc (hunk 1 above)
Comment (psteinfeld) [reply lib/semantics/resolve-names.cc]: Thanks.  I'll make that change.

File: lib/semantics/tools.cc [hunk 4]
```diff
@@ -319,8 +319,6 @@ bool IsTeamType(const DerivedTypeSpec *derived) {
   return IsDerivedTypeFromModule(derived, "iso_fortran_env", "team_type");
 }
 
-bool IsCoarray(const Symbol &symbol) { return symbol.Corank() > 0; }
```
Comment (tskeith) [inline lib/semantics/tools.cc]: This is a better implementation of `IsCoarray()` than the one that replaces it below. For example, this one returns true for a use-associated coarray and the other one returns false.

Comment (psteinfeld) [reply lib/semantics/tools.cc]: Thanks.  I wondered about that.  I'll make that change.

**Pull request merged**